fhir-database-utils/src/main/java/com/ibm/fhir/database/utils/derby/DerbyMaster.java

```diff
@@ -166,75 +213,122 @@ public IDatabaseTranslator getTranslator() {
+            adapter = new DerbyAdapter(printer);
         }
+      
+        // call the Function we've been given using the adapter we just wrapped
+        // around the connection. Each statement executes in its own connection/transaction.
+        fn.accept(adapter);
     }
 
+    /**
+     * Diagnostic utility to display all the current locks in the Derby database
+     */
+    public void dumpLockInfo() {
+        DerbyLockDiag diag = new DerbyLockDiag();
+        IConnectionProvider cp = new DerbyConnectionProvider(this, null);
+        ConnectionProviderTarget target = new ConnectionProviderTarget(cp);
+        DerbyAdapter adapter = new DerbyAdapter(target);
+        List<LockInfo> locks = adapter.runStatement(diag);
+        System.out.println(LockInfo.header());
+        locks.forEach(System.out::println);
+    }
+    
+    /**
+     * Dump locks using the given connection
+     * @param c
+     */
+    public static void dumpLockInfo(Connection c) {
+        // wrap the connection so that we can run our lock diag DAO
+        JdbcTarget target = new JdbcTarget(c);
+        DerbyAdapter adapter = new DerbyAdapter(target);
+        
+        DerbyLockDiag diag = new DerbyLockDiag();
+        List<LockInfo> locks = adapter.runStatement(diag);
+        
+        // render
+        System.out.println(LockInfo.header());
+        locks.forEach(System.out::println);
+    }
+    
+    /**
+     * @implNote this drop is only relevant for in-memory Derby databases, which we no
+     *           longer use due to the size of the tests. This does not remove any
+     *           files on disk.
+     */
     @Override
     public void close() throws Exception {
-        // Drop the database we created
-        boolean dropped = false;
+        shutdown(database);
+    }
+    
+    public static void shutdown(String databaseName) {
+        boolean shutdown = false;
         try {
-            if (connection != null && !connection.isClosed()) {
-                connection.close();
-            }
             Properties properties = new Properties();
             DerbyPropertyAdapter adapter = new DerbyPropertyAdapter(properties);
-            adapter.setDatabase(database);
+            adapter.setDatabase(databaseName);
 
-            final String dropper = DERBY_TRANSLATOR.getUrl(properties) + ";drop=true";
-            logger.info("Dropping derby DB with: " + dropper);
+            final String dropper = DERBY_TRANSLATOR.getUrl(properties) + ";shutdown=true";
+            logger.info("Shutting down Derby DB '" + databaseName + "' with: " + dropper);
+            
+            // should throw an exception if successful
             DriverManager.getConnection(dropper);
         }
         catch (SQLException e) {
-            logger.info("Expected error while closing the database: " + e.getMessage());
-            dropped = true;
+            // should say "Database 'target/derby/...' shutdown."
```

Comment:
I do not follow this message

Reply:
When you shut down a Derby database, it throws an exception to indicate that the shutdown was successful.

Reply:
I think it's a confusing message.